agimus_controller_ros/agimus_controller_ros/agimus_controller.py

```diff
@@ -0,0 +1,241 @@
+        the horizon size and False otherwise.
+        """
+        return len(self.traj_buffer) >= 2 * self.params.ocp.horizon_size
+
+    def first_solve(self, x0: npt.NDArray[np.float64]) -> None:
+        """Initialize OCP, then do the first solve."""
+        # Initialize OCP
+        self.set_weighted_traj_attributes()
+        self.ocp = OCPCrocoHPP(self.rmodel, self.cmodel, self.params.ocp)
+        self.ocp.set_planning_variables(self.x_plan, self.a_plan)
+        self.ocp.set_weights_variables(
+            self.weight_q, self.weight_qdot, self.weight_pose
+        )
+
+        # First solve
+        self.mpc = MPC(self.ocp, self.x_plan, self.a_plan, self.rmodel, self.cmodel)
+        us_init = self.mpc.ocp.u_plan[: self.params.ocp.horizon_size - 1]
+        self.mpc.mpc_first_step(self.x_plan, us_init, x0)
+        self.next_node_idx = self.params.ocp.horizon_size
+        self.get_logger().info("First solve done ")
+
+    def solve(self, x0: npt.NDArray[np.float64]) -> None:
+        """get new trajectory point, reset and solve the OCP problem."""
+        if len(self.traj_buffer) > 0:
+            self.last_point = self.traj_buffer.popleft()
+        self.mpc.mpc_step(x0, self.last_point, self.effector_frame_name)
+
+    def send_control_msg(self, sensor_msg: Sensor) -> None:
+        """Get OCP control output and publish it."""
+        _, tau, K_ricatti = self.mpc.get_mpc_output()
+        ctrl_msg = lfc_py_types.Control(
+            feedback_gain=K_ricatti,
+            feedforward=tau.reshape(self.rmodel.nv, 1),
+            initial_state=sensor_msg,
+        )
+        self.control_publisher.publish(control_numpy_to_msg(ctrl_msg))
+
+    def get_x0_from_sensor_msg(self, sensor_msg: Sensor) -> npt.NDArray[np.float64]:
+        """Create x0 array from robot sensor message."""
+        return np.concatenate(
+            [sensor_msg.joint_state.position, sensor_msg.joint_state.velocity]
+        )
+
+    def run_callback(self, *args) -> None:
+        """
+        Timer callback that checks we can start solve before doing it,
+        then publish messages related to the OCP.
+        """
+        if self.sensor_msg is None:
+            self.get_logger().warn(
+                "Waiting for sensor messages to arrive...",
+                throttle_duration_sec=5.0,
+            )
+            return
+        start_compute_time = time.time()
+        np_sensor_msg: lfc_py_types.Sensor = sensor_msg_to_numpy(self.sensor_msg)
+        if not self.first_run_done:
+            if not self.buffer_has_twice_horizon_points():
+                self.get_logger().warn(
+                    "Waiting for buffer to be filled...",
+                    throttle_duration_sec=5.0,
+                )
+                return
+            self.first_solve(self.get_x0_from_sensor_msg(np_sensor_msg))
+            self.first_run_done = True
+        else:
+            self.solve(self.get_x0_from_sensor_msg(np_sensor_msg))
```

Comment:
```suggestion
        x0 = self.get_x0_from_sensor_msg(np_sensor_msg)
        if not self.first_run_done:
            if not self.buffer_has_twice_horizon_points():
                self.get_logger().warn(
                    "Waiting for buffer to be filled...",
                    throttle_duration_sec=5.0,
                )
                return
            self.first_solve(x0)
            self.first_run_done = True
        else:
            self.solve(x0)
```
IMO this way we avoid copying of the code